### PR TITLE
Issue103inbox

### PR DESCRIPTION
**Pull request by JialinX**

### CODE REVIEW

File: backend/api/models/inbox.py [hunk 1]
```diff
@@ -4,16 +4,17 @@
...
 from api.models.author import Author
 from api.models.post import Post
-from api.models.comment import Comment
+from api.models.follower import Follower
 from api.models.like import Like
 
 MessageTypes = [
     ('post', 'post'),
-    ('comment', 'comment'),
+    ('follow', 'follow'),
```
Comment (AnkushSharma2698) [inline backend/api/models/inbox.py:12]: is messageTypes field even needed now?

File: backend/api/models/inbox.py [hunk 2]
```diff
@@ -4,16 +4,17 @@
...
     ('like', 'like'),
 ]
 
 class Inbox(models.Model):
+
```
Comment (AnkushSharma2698) [inline backend/api/models/inbox.py:17]: how do we know whose inbox this is?

File: backend/api/models/inbox.py [hunk 3]
```diff
@@ -24,14 +25,18 @@ class Inbox(models.Model):
...
     # THIS SHOULD NEVER BE NULL, BUT WE ALSO DONT HAVE A DEFAULT VALUE
     messageType = models.TextField(null=True, choices=MessageTypes) 
+    items = models.JSONField(default=[])
 
     # Not sure how to combine all of the possible foreign keys so I will leave it up to business logic
     # TO be a valid inbox item, it should be checked that at least one of the following id's is not null
-    post_id     = models.ForeignKey(Post, on_delete=models.CASCADE, blank=True, null=True)
-    comment_id  = models.ForeignKey(Comment, on_delete=models.CASCADE, blank=True, null=True)
-    like_id     = models.ForeignKey(Like, on_delete=models.CASCADE, blank=True, null=True)
+    # post_id     = models.ForeignKey(Post, on_delete=models.CASCADE, blank=True, null=True)
```
Comment (AnkushSharma2698) [inline backend/api/models/inbox.py]: if we no longer want these fields pls remove

File: backend/api/services/inboxServices.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    @staticmethod
+    def getInboxByAuthorID(request, author_id):
+        try:
+            all_inbox = inbox.Inbox.objects.filter(author_id=author_id)
+            url = getattr(author.Author.objects.filter(id=author_id)[0], "url")
+            data = {}
+            data["type"] = "inbox"
+            data["author_id"] = url
+            data["tiems"] = []
```
Comment (AnkushSharma2698) [inline backend/api/services/inboxServices.py]: this is a error, should be "items"

File: backend/api/services/inboxServices.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            for each_inbox in all_inbox:
+                item = json.dumps(getattr(each_inbox, "items"))
+                item = json.loads(item)
+                data["tiems"].append(item)
```
Comment (AnkushSharma2698) [inline backend/api/services/inboxServices.py]: here too

File: backend/api/views/inboxView.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+from rest_framework.decorators import api_view
+from rest_framework.response import Response
+from rest_framework import status
+from rest_framework.decorators import permission_classes
+from rest_framework.permissions import AllowAny
+
+from ..services.inboxServices import inboxServices
+
+
+@api_view(['GET'])
```
Comment (AnkushSharma2698) [inline backend/api/views/inboxView.py]: is ths needed?

File: backend/api/views/inboxView.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+def getInbox(request, author_id):
+    res = inboxServices.getInboxByAuthorID(request, author_id)
+    return res
+
+@api_view(['POST'])
+def sendToInbox(request, author_id):
```
Comment (AnkushSharma2698) [inline backend/api/views/inboxView.py]: this one

File: backend/api/views/inboxView.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    res = inboxServices.sendInboxByAuthorID(request, author_id)
+    return res
+
+@api_view(['DELETE'])
+def clearInbox(request, author_id):
+    res = inboxServices.clearInboxByAuthorID(request, author_id)
```
Comment (AnkushSharma2698) [inline backend/api/views/inboxView.py]: is this needed?

File: backend/api/services/inboxServices.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        body = json.loads(body)
+        request_data = {}
+        
+        request_data['author_id'] = author_id
+        items = body["items"]
+        try:
+            for item in items:
+                type = item["type"]
+                request_data['messageType'] = type
+                if type in ["post","follow","like"]:
```
Comment (AnkushSharma2698) [inline backend/api/services/inboxServices.py:41]: the API for type `like` is actually `Like` if u look at the spec

File: backend/api/models/inbox.py [hunk 10]
```diff
@@ -4,16 +4,17 @@
...
 from api.models.post import Post
-from api.models.comment import Comment
+from api.models.follower import Follower
 from api.models.like import Like
 
 MessageTypes = [
     ('post', 'post'),
-    ('comment', 'comment'),
+    ('follow', 'follow'),
     ('like', 'like'),
```
Comment (AnkushSharma2698) [inline backend/api/models/inbox.py:13]: This must become `('Like', 'Like')`

File: backend/api/services/inboxServices.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                    request_data["items"] = item
+                else:
+                    return Response(status=status.HTTP_400_BAD_REQUEST)
+                serializer = InboxSerializer(data=request_data)
+                if serializer.is_valid():
+                    instance = serializer.save()
+                else:
+                    return Response(status=status.HTTP_400_BAD_REQUEST)
```
Comment (AnkushSharma2698) [inline backend/api/services/inboxServices.py:49]: provide error message

File: backend/api/services/inboxServices.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+                instance.save()
+
+            return Response(status=status.HTTP_201_CREATED)
+        except:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
```
Comment (AnkushSharma2698) [inline backend/api/services/inboxServices.py:54]: provide error message

File: backend/api/services/inboxServices.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+            data["type"] = "inbox"
+            data["author_id"] = url
+            data["items"] = []
+            for each_inbox in all_inbox:
+                item = json.dumps(getattr(each_inbox, "items"))
+                item = json.loads(item)
+                data["items"].append(item)
+            return Response(data)
+        except:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
```
Comment (AnkushSharma2698) [inline backend/api/services/inboxServices.py:26]: provide error message

File: backend/api/services/inboxServices.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+    @staticmethod
+    def clearInboxByAuthorID(request, author_id):
+        try:
+            data = inbox.Inbox.objects.filter(author_id=author_id).delete()
+            if data[0]:
+                return Response(status=status.HTTP_200_OK)
+            # author id is not found
+            else:
+                return Response(status=status.HTTP_404_NOT_FOUND)
```
Comment (AnkushSharma2698) [inline backend/api/services/inboxServices.py:64]: provide error message

**Comment by JialinX** [comment]
Merged by pr#134

**Pull request closed without merging**